src/transform/deadCode.ml

```diff
@@ -0,0 +1,70 @@
+open Batteries
+open GoblintCil
+
+
+(** Filter statements out of a block (recursively). CFG fields (prev, next, etc.) are no longer valid after calling.
+    Returns true if anything is left in block, false if the block is now empty.
+    Invariants:
+    - f (goto label) ==> f (labelled stmt), i.e. if a goto statement is not filtered out, the target may not be filtered out either.
+    - block may not contain switch statements *)
+let filter_map_block f (block : Cil.block) : bool =
```

Comment:
Could gotos be spuriously detected as live.

Reply:
Did you come to a conclusion on this?

Reply:
The case that would cause a problem is if a `goto label;` remains in the block, but `label: ...` is removed. That means the goto must be either (1) live in some context or (2) not appear in the result at all. The labelled statement on the other hand must be dead in all contexts.

In case (1), the goto is live in some context and control flow always passes to the labelled statement, so the labelled statement would also need to be live.

In case (2), the goto would not have a CFG node in the result. This case only occurs for statements without a node in the result, e.g. compound statements. As far as I understand, this case can not occur.